### ISSUE TITLE
`reference/test/solver/bicgstab_kernels.cpp` tests fail when converted to `float` (i.e., single precision)

### ISSUE THREAD

**Issue by venovako** [ginkgo-project/ginkgo#38]
Having converted `reference/test/solver/bicgstab_kernels.cpp` to use single precision by copying it to `reference/test/solver/bicgstab_kernels_float.cpp` and making changes as per diff below:
```diff
51c51
<     using Mtx = gko::matrix::Dense<>;
---
>     using Mtx = gko::matrix::Dense<float>;
57c57
<               gko::solver::BicgstabFactory<>::create(exec, 8, 1e-15))
---
>               gko::solver::BicgstabFactory<float>::create(exec, 8, 1e-7))
62c62
<     std::unique_ptr<gko::solver::BicgstabFactory<>> bicgstab_factory;
---
>     std::unique_ptr<gko::solver::BicgstabFactory<float>> bicgstab_factory;
74c74
<     ASSERT_MTX_NEAR(x, l({-4.0, -1.0, 4.0}), 1e-8);
---
>     ASSERT_MTX_NEAR(x, l({-4.0, -1.0, 4.0}), 1e-4);
87c87
<     ASSERT_MTX_NEAR(x, l({{-4.0, 1.0}, {-1.0, 2.0}, {4.0, -1.0}}), 1e-8);
---
>     ASSERT_MTX_NEAR(x, l({{-4.0, 1.0}, {-1.0, 2.0}, {4.0, -1.0}}), 1e-4);
102c102
<     ASSERT_MTX_NEAR(x, l({-8.5, -3.0, 6.0}), 1e-8);
---
>     ASSERT_MTX_NEAR(x, l({-8.5, -3.0, 6.0}), 1e-4);
118c118
<     ASSERT_MTX_NEAR(x, l({{-8.5, 1.0}, {-3.0, 2.0}, {6.0, -5.0}}), 1e-8);
---
>     ASSERT_MTX_NEAR(x, l({{-8.5, 1.0}, {-3.0, 2.0}, {6.0, -5.0}}), 1e-4);
```
the following tests fail by not attaining the required accuracy:
```
[ RUN      ] Bicgstab.SolvesMultipleDenseSystems
.../ginkgo/reference/test/solver/bicgstab_kernels_float.cpp:87: Failure
Relative error between x and {{-4.0, 1.0}, {-1.0, 2.0}, {4.0, -1.0}} is 0.11072188922929171
	which is larger than 1e-4 (which is 0.0001)
x is:
	-4.0000228881835938	0.54083889722824097	
	-1.0000052452087402	1.858891487121582	
	4.000030517578125	-0.50263333320617676	
{{-4.0, 1.0}, {-1.0, 2.0}, {4.0, -1.0}} is:
	-4	1	
	-1	2	
	4	-1	
component-wise relative error is:
	5.7220131568155855e-06	0.45916110277175903	
	5.2451812281639528e-06	0.070554256439208984	
	7.6293363240331724e-06	0.49736666679382324	

[  FAILED  ] Bicgstab.SolvesMultipleDenseSystems (1 ms)
[ RUN      ] Bicgstab.SolvesDenseSystemUsingAdvancedApply
.../ginkgo/reference/test/solver/bicgstab_kernels_float.cpp:102: Failure
Relative error between x and {-8.5, -3.0, 6.0} is 0.0014036462767795599
	which is larger than 1e-4 (which is 0.0001)
x is:
	-8.4900836944580078	
	-2.9963669776916504	
	5.9890694618225098	
{-8.5, -3.0, 6.0} is:
	-8.5	
	-3	
	6	
component-wise relative error is:
	0.0011666241814108455	
	0.0012110074361165364	
	0.0018217563629150391	

[  FAILED  ] Bicgstab.SolvesDenseSystemUsingAdvancedApply (0 ms)
[ RUN      ] Bicgstab.SolvesMultipleDenseSystemsUsingAdvancedApply
.../ginkgo/reference/test/solver/bicgstab_kernels_float.cpp:118: Failure
Relative error between x and {{-8.5, 1.0}, {-3.0, 2.0}, {6.0, -5.0}} is 0.0015665962953109148
	which is larger than 1e-4 (which is 0.0001)
x is:
	-8.4900836944580078	0.99250507354736328	
	-2.9963669776916504	1.9976270198822021	
	5.9890694618225098	-4.9917192459106445	
{{-8.5, 1.0}, {-3.0, 2.0}, {6.0, -5.0}} is:
	-8.5	1	
	-3	2	
	6	-5	
component-wise relative error is:
	0.0011666241814108455	0.0074949264526367188	
	0.0012110074361165364	0.0011864900588989258	
	0.0018217563629150391	0.0016561508178710937	

[  FAILED  ] Bicgstab.SolvesMultipleDenseSystemsUsingAdvancedApply (1 ms)
```

**Comment by gflegar** [comment]
The problem here could be one of the following:

*   There's something wrong with the implementation of BiCGStab.
*   There's a problem with how we test. The number `1e-15` means "reduce the original residual by 15 orders of magnitude", not "obtain accuracy of 1e-15". Thus, if the original residual was to large (which we don't test anywhere), then the single precision version could might as well reduce the original residual by 7 orders of magnitude, and still get a solution which only has 1-3 correct digits. I think it would be more useful if we use a hard limit on residual norm for tests, but we have to implement different stopping criteria to do that.

**Comment by gflegar** [comment]
Btw, there's an issue on the private repo about implementing tese stopping criteria already: https://gitlab.com/ginkgo-project/ginkgo/issues/60 Maybe we should consider moving it into the public repo.

**Comment by venovako** [comment]
Sorry if I misunderstood the parameters.
Please also note that the test with one RHS (`SolvesDenseSystem`) succeeds in `float`, if that may indicate you something.

**Comment by gflegar** [comment]
> Sorry if I misunderstood the parameters.

Even if you did, it's not an issue here. There's still something wrong with the tests.



> Please also note that the test with one RHS (SolvesDenseSystem) succeeds in float, if that may indicate you something.

I noticed, and the explanation is probably this: https://gitlab.com/ginkgo-project/ginkgo/issues/55
TL; DR The convergence test is not exactly the same for multiple RHS as is for one RHS. Another thing we have to fix.

**Comment by tcojean** [comment]
Yes, actually I remember now that we saw a similar problem for double when I integrated that code, and we had to settle with the current parameters for it to work. But we need to tackle the related issue in the private repo to make this work properly in the general case.

**Comment by gflegar** [comment]
@tcojean, @hartwiganzt now that we have stopping criteria - maybe we should take another look into this?

**Comment by tcojean** [comment]
I just checked this again. It looks like we have now fixed this problem in the meantime.

Result:
```
$ ./reference/test/solver/bicgstab_kernels
Running main() from /home/tcojean/ginkgo-github/build/third_party/gtest/src/googletest/src/gtest_main.cc
[==========] Running 7 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 7 tests from Bicgstab
[ RUN      ] Bicgstab.SolvesDenseSystem
[       OK ] Bicgstab.SolvesDenseSystem (0 ms)
[ RUN      ] Bicgstab.SolvesMultipleDenseSystems
[       OK ] Bicgstab.SolvesMultipleDenseSystems (0 ms)
[ RUN      ] Bicgstab.SolvesDenseSystemUsingAdvancedApply
[       OK ] Bicgstab.SolvesDenseSystemUsingAdvancedApply (1 ms)
[ RUN      ] Bicgstab.SolvesMultipleDenseSystemsUsingAdvancedApply
[       OK ] Bicgstab.SolvesMultipleDenseSystemsUsingAdvancedApply (0 ms)
[ RUN      ] Bicgstab.SolvesBigDenseSystemForDivergenceCheck1
[       OK ] Bicgstab.SolvesBigDenseSystemForDivergenceCheck1 (0 ms)
[ RUN      ] Bicgstab.SolvesBigDenseSystemForDivergenceCheck2
[       OK ] Bicgstab.SolvesBigDenseSystemForDivergenceCheck2 (1 ms)
[ RUN      ] Bicgstab.SolvesMultipleDenseSystemsDivergenceCheck
[       OK ] Bicgstab.SolvesMultipleDenseSystemsDivergenceCheck (0 ms)
[----------] 7 tests from Bicgstab (2 ms total)

[----------] Global test environment tear-down
[==========] 7 tests from 1 test case ran. (3 ms total)
[  PASSED  ] 7 tests.
```

Patch I used:
```diff
diff --git a/reference/test/solver/bicgstab_kernels.cpp b/reference/test/solver/bicgstab_kernels.cpp
index bb89d71f..bf37f9b0 100644
--- a/reference/test/solver/bicgstab_kernels.cpp
+++ b/reference/test/solver/bicgstab_kernels.cpp
@@ -52,8 +52,8 @@ namespace {

 class Bicgstab : public ::testing::Test {
 protected:
-    using Mtx = gko::matrix::Dense<>;
-    using Solver = gko::solver::Bicgstab<>;
+    using Mtx = gko::matrix::Dense<float>;
+    using Solver = gko::solver::Bicgstab<float>;

     Bicgstab()
         : exec(gko::ReferenceExecutor::create()),
@@ -66,28 +66,28 @@ protected:
                       gko::stop::Time::build()
                           .with_time_limit(std::chrono::seconds(6))
                           .on(exec),
-                      gko::stop::ResidualNormReduction<>::build()
-                          .with_reduction_factor(1e-15)
+                      gko::stop::ResidualNormReduction<float>::build()
+                          .with_reduction_factor(1e-7)
                           .on(exec))
                   .on(exec)),
           bicgstab_factory_precision(
-              gko::solver::Bicgstab<>::build()
+              gko::solver::Bicgstab<float>::build()
                   .with_criteria(
                       gko::stop::Iteration::build().with_max_iters(50u).on(
                           exec),
                       gko::stop::Time::build()
                           .with_time_limit(std::chrono::seconds(6))
                           .on(exec),
-                      gko::stop::ResidualNormReduction<>::build()
-                          .with_reduction_factor(1e-15)
+                      gko::stop::ResidualNormReduction<float>::build()
+                          .with_reduction_factor(1e-7)
                           .on(exec))
                   .on(exec))
     {}

     std::shared_ptr<const gko::Executor> exec;
     std::shared_ptr<Mtx> mtx;
-    std::unique_ptr<gko::solver::Bicgstab<>::Factory> bicgstab_factory;
-    std::unique_ptr<gko::solver::Bicgstab<>::Factory>
+    std::unique_ptr<gko::solver::Bicgstab<float>::Factory> bicgstab_factory;
+    std::unique_ptr<gko::solver::Bicgstab<float>::Factory>
         bicgstab_factory_precision;
 };

@@ -100,7 +100,7 @@ TEST_F(Bicgstab, SolvesDenseSystem)

     solver->apply(b.get(), x.get());

-    ASSERT_MTX_NEAR(x, l({-4.0, -1.0, 4.0}), 1e-8);
+    ASSERT_MTX_NEAR(x, l({-4.0, -1.0, 4.0}), 1e-4);
 }


@@ -113,7 +113,7 @@ TEST_F(Bicgstab, SolvesMultipleDenseSystems)

     solver->apply(b.get(), x.get());

-    ASSERT_MTX_NEAR(x, l({{-4.0, 1.0}, {-1.0, 2.0}, {4.0, -1.0}}), 1e-8);
+    ASSERT_MTX_NEAR(x, l({{-4.0, 1.0}, {-1.0, 2.0}, {4.0, -1.0}}), 1e-4);
 }


@@ -128,7 +128,7 @@ TEST_F(Bicgstab, SolvesDenseSystemUsingAdvancedApply)
     solver->apply(alpha.get(), b.get(), beta.get(), x.get());


-    ASSERT_MTX_NEAR(x, l({-8.5, -3.0, 6.0}), 1e-8);
+    ASSERT_MTX_NEAR(x, l({-8.5, -3.0, 6.0}), 1e-4);
 }


@@ -144,7 +144,7 @@ TEST_F(Bicgstab, SolvesMultipleDenseSystemsUsingAdvancedApply)
     solver->apply(alpha.get(), b.get(), beta.get(), x.get());


-    ASSERT_MTX_NEAR(x, l({{-8.5, 1.0}, {-3.0, 2.0}, {6.0, -5.0}}), 1e-8);
+    ASSERT_MTX_NEAR(x, l({{-8.5, 1.0}, {-3.0, 2.0}, {6.0, -5.0}}), 1e-4);
 }


@@ -169,7 +169,7 @@ TEST_F(Bicgstab, SolvesBigDenseSystemForDivergenceCheck1)
         x,
         l({0.13853406350816114, -0.08147485210505287, -0.0450299311807042,
            -0.0051264177562865719, 0.11609654300797841, 0.1018688746740561}),
-        1e-9);
+        1e-4);
 }


@@ -193,16 +193,17 @@ TEST_F(Bicgstab, SolvesBigDenseSystemForDivergenceCheck2)
         x,
         l({0.13517641417299162, 0.75117689075221139, 0.47572853185155239,
            -0.50927993095367852, 0.13463333820848167, 0.23126768306576015}),
-        1e-9);
+        1e-4);
 }


-double infNorm(gko::matrix::Dense<> *mat, size_t col = 0)
+template <typename T>
+T infNorm(gko::matrix::Dense<T> *mat, size_t col = 0)
 {
     using std::abs;
-    double norm = 0.0;
+    T norm = 0.0;
     for (size_t i = 0; i < mat->get_size()[0]; ++i) {
-        double absEntry = abs(mat->at(i, col));
+        T absEntry = abs(mat->at(i, col));
         if (norm < absEntry) norm = absEntry;
     }
     return norm;
@@ -273,7 +274,7 @@ TEST_F(Bicgstab, SolvesMultipleDenseSystemsDivergenceCheck)

     // Not sure if this is necessary, the assertions above should cover what is
     // needed.
-    ASSERT_MTX_NEAR(xc, testMtx, 1e-14);
+    ASSERT_MTX_NEAR(xc, testMtx, 1e-4);
 }
```
